superset-frontend/src/components/DropdownContainer/index.tsx

```diff
@@ -0,0 +1,233 @@
+   */
+  popoverTriggerText?: string;
+  /**
+   * Main container additional style properties.
+   */
+  style?: CSSProperties;
+}
+
+const DropdownContainer = ({
+  items,
+  onOverflowingStateChange,
+  popoverStyle = {},
+  popoverTriggerIcon,
+  popoverTriggerText = t('More'),
+  style,
+}: DropdownContainerProps) => {
+  const theme = useTheme();
+  const { ref, width = 0 } = useResizeDetector<HTMLDivElement>();
+  const previousWidth = usePrevious(width) || 0;
+  const { current } = ref;
+  const [overflowingIndex, setOverflowingIndex] = useState<number>(-1);
+  const [itemsWidth, setItemsWidth] = useState<number[]>([]);
+
+  useEffect(() => {
+    const container = current?.children.item(0);
+    if (container) {
+      const { children } = container;
+      const childrenArray = Array.from(children);
+
+      // Stores items width once
+      if (itemsWidth.length === 0) {
+        setItemsWidth(
+          childrenArray.map(child => child.getBoundingClientRect().width),
+        );
+      }
+
+      // Calculates the index of the first overflowed element
+      const index = childrenArray.findIndex(
+        child =>
+          child.getBoundingClientRect().right >
+          container.getBoundingClientRect().right,
+      );
+      setOverflowingIndex(index === -1 ? children.length : index);
+
+      if (width > previousWidth && overflowingIndex !== -1) {
+        // Calculates remaining space in the container
+        const button = current?.children.item(1);
+        const buttonRight = button?.getBoundingClientRect().right || 0;
+        const containerRight = current?.getBoundingClientRect().right || 0;
+        const remainingSpace = containerRight - buttonRight;
+        // Checks if the first element in the popover fits in the remaining space
+        const fitsInRemainingSpace = remainingSpace >= itemsWidth[0];
+        if (fitsInRemainingSpace && overflowingIndex < items.length) {
+          // Moves element from popover to container
+          setOverflowingIndex(overflowingIndex + 1);
+        }
+      }
+    }
+  }, [
+    current,
+    items.length,
+    itemsWidth,
+    overflowingIndex,
+    previousWidth,
+    width,
+  ]);
+
+  const notOverflowedItems = useMemo(
+    () =>
+      items
+        .slice(0, overflowingIndex !== -1 ? overflowingIndex : items.length)
+        .map(item => React.cloneElement(item.element, { key: item.id })),
+    [items, overflowingIndex],
+  );
+
+  const notOverflowedIds = useMemo(
+    () =>
+      items
+        .slice(0, overflowingIndex !== -1 ? overflowingIndex : items.length)
+        .map(item => item.id),
+    [items, overflowingIndex],
+  );
+
+  const overflowedItems = useMemo(
+    () =>
+      overflowingIndex !== -1
+        ? items
+            .slice(overflowingIndex, items.length)
+            .map(item => React.cloneElement(item.element, { key: item.id }))
+        : [],
+    [items, overflowingIndex],
+  );
+
+  const overflowedIds = useMemo(
```

Comment:
Could we maybe merge calculating items and ids in 1 operation? Like `const [overflowedItems, overflowedIds] = useMemo(...)`

Reply:
Done 👍🏼